### PR TITLE
Service accounts update grpc public handler

### PR DESCRIPTION
**Pull request by cesnietor**
This pr allows to update a service account by single fields (name, enabled) and all it's corresponding permissions assigned to it.
Update for service accounts is like: PUT `/api/v1/service_accounts/{id}`
```
{
"name": "new name for asdSA",
"enabled": true,
"permission_ids": ["67e6b0fa-a24e-484d-99b3-828f7516eee9"]
}
```
and returns :
```
{
    "service_account": {
        "id": "be6774a9-72d4-4f21-9125-345d4560f2b0",
        "name": "new name for asdSA",
        "access_key": "DNDE55LTM53PVATE",
        "enabled": true
    },
    "permissions": [
        {
            "id": "67e6b0fa-a24e-484d-99b3-828f7516eee9",
            "name": "new permission greate name",
            "slug": "permission-greate",
            "description": "this is the second description",
            "effect": "Deny",
            "resources": [
                {
                    "id": "3c9d8e98-0874-4965-b23b-3a6973fd7f89",
                    "bucket_name": "cesnietor",
                    "pattern": "*"
                }
            ],
            "actions": [
                {
                    "id": "8c3b7171-1163-4095-96b0-d59a34930b01",
                    "type": "read"
                },
                {
                    "id": "60694e5e-52fa-4574-9b5c-361ed2cd7240",
                    "type": "write"
                }
            ]
        }
    ]
}
```

### CODE REVIEW

File: api/api-public-tenant-service-account.go [hunk 1]
```diff
@@ -117,6 +120,134 @@ func (s *server) ListServiceAccounts(ctx context.Context, in *pb.ListServiceAcco
...
+	if idRequest == "" {
+		return nil, status.New(codes.InvalidArgument, "an id is needed").Err()
+	}
+	if nameRequest == "" {
+		return nil, status.New(codes.InvalidArgument, "an name is needed").Err()
+	}
+	if len(permisionsIDs) == 0 {
+		return nil, status.New(codes.InvalidArgument, "a list of permissions is needed").Err()
+	}
+	// start app context
```
Comment (dvaldivia) [inline api/api-public-tenant-service-account.go:137]: this should error out immediately if they try to update the access_key

File: api/api-public-tenant-service-account.go [hunk 2]
```diff
@@ -117,6 +120,134 @@ func (s *server) ListServiceAccounts(ctx context.Context, in *pb.ListServiceAcco
...
+
+	defer func() {
+		if err != nil {
+			appCtx.Rollback()
+			return
+		}
+	}()
+
+	// Fetch the service account
+	id, err := uuid.FromString(idRequest)
```
Comment (dvaldivia) [inline api/api-public-tenant-service-account.go:151]: can you put all this logic inside a re-usable function so I can expose this functionality to the CLI?

File: api/api-public-tenant-service-account.go [hunk 3]
```diff
@@ -117,6 +120,134 @@ func (s *server) ListServiceAccounts(ctx context.Context, in *pb.ListServiceAcco
...
+		return nil, status.New(codes.Internal, "Internal error").Err()
+	}
+
+	// Compare current Permissions with the desired ones
+
+	var currentPerms []string
+	for _, perm := range perms {
+		currentPerms = append(currentPerms, perm.ID.String())
+	}
+	fmt.Println("currentPerms: ", currentPerms)
```
Comment (dvaldivia) [inline api/api-public-tenant-service-account.go]: maybe not needed

File: api/api-public-tenant-service-account.go [hunk 4]
```diff
@@ -117,6 +120,134 @@ func (s *server) ListServiceAccounts(ctx context.Context, in *pb.ListServiceAcco
...
+	permissionsToCreate := differenceArrays(permisionsIDs, currentPerms)
+	permissionsToDelete := differenceArrays(currentPerms, permisionsIDs)
+
+	// Create new service_accounts_permissions
+	permsToCreateIDs, err := cluster.UUIDsFromStringArr(permissionsToCreate)
+	if err != nil {
+		log.Println(err.Error())
+		return nil, status.New(codes.InvalidArgument, "invalid permission id").Err()
+	}
+	fmt.Println("tocreate: ", permsToCreateIDs)
```
Comment (dvaldivia) [inline api/api-public-tenant-service-account.go]: maybe not needed

File: api/api-public-tenant-service-account.go [hunk 5]
```diff
@@ -117,6 +120,134 @@ func (s *server) ListServiceAccounts(ctx context.Context, in *pb.ListServiceAcco
...
+	if err != nil {
+		log.Println(err.Error())
+		return nil, status.New(codes.Internal, "Internal error").Err()
+	}
+	permsToDeleteIDs, err := cluster.UUIDsFromStringArr(permissionsToDelete)
+	if err != nil {
+		log.Println(err.Error())
+		return nil, status.New(codes.InvalidArgument, "invalid permission id").Err()
+	}
+	fmt.Println("todelete: ", permsToDeleteIDs)
```
Comment (dvaldivia) [inline api/api-public-tenant-service-account.go]: maybe not needed

File: cluster/permissions.go [hunk 6]
```diff
@@ -601,6 +627,53 @@ func assignPermissionToMultipleSAsOnDB(ctx *Context, permission *uuid.UUID, serv
 	return nil
 }
 
+// AssignMultiplePermissionsToSADB inserts on table service_accounts_permissions, multiple permissions to a single service account
+func AssignMultiplePermissionsToSADB(ctx *Context, serviceAccountID *uuid.UUID, permissionsIDs []*uuid.UUID) error {
```
Comment (dvaldivia) [inline cluster/permissions.go:631]: SABDASKDJHASD

File: cluster/permissions.go [hunk 7]
```diff
@@ -601,6 +627,53 @@ func assignPermissionToMultipleSAsOnDB(ctx *Context, permission *uuid.UUID, serv
...
+		_, err := stmt.Exec(serviceAccountID, permissionID, ctx.WhoAmI)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// DeleteMultiplePermissionsOnSADB removes on table service_accounts_permissions, multiple permissions on a single service account
+func DeleteMultiplePermissionsOnSADB(ctx *Context, serviceAccountID *uuid.UUID, permissionsIDs []*uuid.UUID) error {
```
Comment (dvaldivia) [inline cluster/permissions.go:658]: SABA DABA DA 
SABA DABA DA 
En el centro del planeta
Shabadabada, shabadabada
Olvidado en la banqueta
Shabadabada, shabadabada
En un cráter en la luna
Shabadabada, shabadabada
En el trueno o en la lluvia

**Review by dvaldivia — CHANGES_REQUESTED** [review]

File: api/api-public-tenant-service-account.go (hunk 1 above)
Comment (cesnietor) [reply api/api-public-tenant-service-account.go:137]: access_key is not defined on the proto as a Request so we don't care if they send it.

File: api/api-public-tenant-service-account.go (hunk 3 above)
Comment (cesnietor) [reply api/api-public-tenant-service-account.go]: done

File: api/api-public-tenant-service-account.go (hunk 5 above)
Comment (cesnietor) [reply api/api-public-tenant-service-account.go]: done

File: api/api-public-tenant-service-account.go (hunk 2 above)
Comment (cesnietor) [reply api/api-public-tenant-service-account.go:151]: done

Comment (cesnietor) [reply api/api-public-tenant-service-account.go:151]: `UpdateServiceAccountFields()` was created

**Review by dvaldivia — APPROVED** [review]

**Pull request merged**